File: ui/service/FeatureService.ts
import { providerStore } from '../../ui';

export const prepareFeatureString = (feature: string): string => {
  return feature.toLowerCase().trim();
};

export class FeatureService {
  public isEnabled(feature: string): boolean {
    const mapState = providerStore.getState() as any;
    const { config } = mapState.toJS();
    const { featureFlags } = config;
    return featureFlags
      .map((featureFlag: string) => prepareFeatureString(featureFlag))
      .includes(prepareFeatureString(feature));
  }

  public all(): void {
    const mapState = providerStore.getState() as any;
    return mapState.toJS().config.featureFlag;
  }
}

File: common/src/main/scala/com/heimdali/services/FeatureService.scala
package com.heimdali.services

import com.typesafe.scalalogging.LazyLogging

trait FeatureService[F[_]] {

  def isEnabled(feature: String): Boolean

  def runIfEnabled[T](feature: String, action: => T, alternative: => T): T

  def all(): List[String]
}

class FeatureServiceImpl[F[_]](featureFlags: String) extends FeatureService[F] with LazyLogging {

  private val flagsList = featureFlags.split(",").toList

  override def isEnabled(feature: String): Boolean = {
    logger.info(s"Enabled feature flags: $featureFlags")
    logger.debug(s"Checking is $feature is enabled")
    flagsList.exists(flag => flag.trim.toLowerCase == feature.trim.toLowerCase)
  }

  override def runIfEnabled[T](feature: String, action: => T, alternative: => T): T = {
    if (isEnabled(feature)) action else alternative
  }

  override def all(): List[String] = flagsList
}
